[Complex Text Layouts] Implement TextServer interface.

bruvzg opened the pull request:
Part of https://github.com/godotengine/godot-proposals/issues/1180 implementation.

~Depends on #40999~

#### TODO:

- TextServer interface. ✔️
   - TextServer Fallback implementation (based on existing Font implementations) ✔️
      - Dynamic font data ✔️
      - Bitmap font data ✔️
   - TextServer ICU + HarfBuzz implementation. ✔️
      - Dynamic font data ✔️
      - Bitmap font data ✔️
   - TextServer documentation ✔️
   - TextServer unit test + test project ✔️
- Font class refactoring ✔️
- Simple controls (everything except `Label`, `LineEdit`, `TextEdit`, `RichTextLabel`) TS base text ✔️
- Interface mirroring ✔️
- Add BiDi override for path `LineEdit`s and `CodeEdit` ✔️
- Complex controls ✔️
   - `Label` ✔️
   - `LineEdit` ✔️
   - `TextEdit` + `CodeEdit` ✔️
   - `RichTextLabel` (probably will be done as the separate PR)
- Editor tools for Font substitution config and TS data export ✔️
- Update controls / font documentation ✔️
- Bump `MessageQuery` size to 2048 KB to handle additional `LineEdit` / `TextEdit` popup submenus (superseded by #42296) ✔️

#### Breaking changes:

- `BitmapFont`, `DynamicFont` and `DyamicFontData` are removed and replaced with universal `Font` + `FontData`, actual implementation is handled by `TextServer`.
- Font size and outline parameters are no longer font properties and can be passed to draw functions directly, theme properties added accordingly.
- `Font` and `CanvasItem` API changes:
  - Added `draw_multiline_string` function.
  - Added additional parameters to all draw functions.
  - New font fallback API similar to `DynamicFont` fallback (single list of `FontData`).

<details>
  <summary>Major Font API changes:</summary>
Old:

```c++
	Size2 get_string_size(const String &p_string) const;
	void draw(RID p_canvas_item, const Point2 &p_pos, const String &p_text, const Color &p_modulate = Color(1, 1, 1), int p_clip_w = -1, const Color &p_outline_modulate = Color(1, 1, 1)) const;

	void draw_halign(RID p_canvas_item, const Point2 &p_pos, HAlign p_align, float p_width, const String &p_text, const Color &p_modulate = Color(1, 1, 1), const Color &p_outline_modulate = Color(1, 1, 1)) const;

	Size2 get_wordwrap_string_size(const String &p_string, float p_width) const;

	Size2 get_char_size(char32_t p_char, char32_t p_next = 0) const;
	float draw_char(RID p_canvas_item, const Point2 &p_pos, char32_t p_char, char32_t p_next = 0, const Color &p_modulate = Color(1, 1, 1), bool p_outline = false) const;
```

New:

```c++
	Size2 get_string_size(const String &p_text, int p_size = -1) const;
	void draw_string(RID p_canvas_item, const Point2 &p_pos, const String &p_text, HAlign p_align = HALIGN_LEFT, float p_width = -1, int p_size = -1, const Color &p_modulate = Color(1, 1, 1), int p_outline_size = 0, const Color &p_outline_modulate = Color(1, 1, 1, 0), uint8_t p_flags = TextServer::JUSTIFICATION_KASHIDA | TextServer::JUSTIFICATION_WORD_BOUND) const;

	Size2 get_multiline_string_size(const String &p_text, float p_width = -1, int p_size = -1, uint8_t p_flags = TextServer::BREAK_MANDATORY | TextServer::BREAK_WORD_BOUND) const;
	void draw_multiline_string(RID p_canvas_item, const Point2 &p_pos, const String &p_text, HAlign p_align = HALIGN_LEFT, float p_width = -1, int p_max_lines = -1, int p_size = -1, const Color &p_modulate = Color(1, 1, 1), int p_outline_size = 0, const Color &p_outline_modulate = Color(1, 1, 1, 0), uint8_t p_flags = TextServer::BREAK_MANDATORY | TextServer::BREAK_WORD_BOUND | TextServer::JUSTIFICATION_KASHIDA | TextServer::JUSTIFICATION_WORD_BOUND) const;

	Size2 get_char_size(char32_t p_char, char32_t p_next = 0, int p_size = -1) const;
	float draw_char(RID p_canvas_item, const Point2 &p_pos, char32_t p_char, char32_t p_next = 0, int p_size = -1, const Color &p_modulate = Color(1, 1, 1), int p_outline_size = 0, const Color &p_outline_modulate = Color(1, 1, 1, 0)) const;
```

</details>

#### TextServer interface structure:

![ts_map](https://user-images.githubusercontent.com/7645683/92443294-fb904000-f1b9-11ea-806a-0dcc61b17cda.png)

#### Test projects:

GDNative plugin template: https://github.com/bruvzg/gdnative_text_server_template

#### Current status:
- Backend: Working.
- Controls: All controls except `RichTextLabel` and containers support mirroring and use `TexeServer` base text rendering.

<img width="956" alt="Screenshot 2020-09-10 at 15 33 07" src="https://user-images.githubusercontent.com/7645683/92729751-781e4c80-f37b-11ea-903a-1e1487a03823.png">

<img width="1720" alt="Screenshot 2020-10-29 at 13 37 53" src="https://user-images.githubusercontent.com/7645683/97563749-2ba2d500-19ec-11eb-99b5-0a73b3407568.png">

### Input:
Input of the characters outside BMP (codes `> 0xFFFF`, emoji, rare/ancient scripts, e.t.c.) is fully tested on macOS, Windows 10, Android and Wine under macOS.

Should work on Linux (input is received as UTF-8 string), but might need some more testing with different input methods.

On iOS, it should work as well (input is received as UTF-16 string), but I can't test it since I do not have any iOS device, and simulator currently do not have all Metal features supported to run Godot.

<details>
  <summary>Screenshots - click to expand</summary>

![android](https://user-images.githubusercontent.com/7645683/98096894-ef142500-1e94-11eb-92dd-00e9f26f8ae2.jpg)
![macos](https://user-images.githubusercontent.com/7645683/98096904-f2a7ac00-1e94-11eb-82a6-9c4610cb0b2f.png)
![ubuntu](https://user-images.githubusercontent.com/7645683/98096907-f3d8d900-1e94-11eb-824f-cf242b4096a9.png)
![win_10](https://user-images.githubusercontent.com/7645683/98096911-f4716f80-1e94-11eb-9d73-311d8e2d3886.png)

</details>

### Fixed and related issues / PRs:

Previous implementations: #21791, #10546

Fixes #40688
Fixes #28371
Fixes #3081
Fixes #9961
Fixes #982

bruvzg committed:
[Complex Text Layouts] Implement TextServer interface. Implement Fallback TextServer.

bruvzg committed:
[Complex Text Layouts] Add third-party TextServer dependencies (ICU, HarfBuzz, Graphite).

bruvzg committed:
[Complex Text Layouts] Implement ICU / HarfBuzz based TextServer module.

bruvzg committed:
[Complex Text Layouts] Implement GDNative interface for TextServer.

bruvzg committed:
[Complex Text Layouts] Add TextServer documentation. Update Font, CanvasItem, Theme and modified controls documentation.

bruvzg committed:
[Complex Text Layouts] Refactor Font class, default themes and controls to use Text Server interface.
Implement interface mirroring.
Add TextLine and TextParagraph classes.
Handle UTF-16 input on macOS and Windows.

bruvzg committed:
[Complex Text Layouts] Refactor Label and LineEdit controls.

bruvzg committed:
[Complex Text Layouts] Refactor TextEdit and CodeEdit controls.

bruvzg commented:
Everything seems to be working now (except for the RichTextLabel, which will be done as separate PR), but probably I have missed some glitches.

List of changes in the PR:
- Added `TextServerManager` (control loading, enumeration and switching for `TextServer`s).
- Added `TextServer` interface (handles fonts, text rendering and line breaking and input related stuff).
  - Handles:
    - BiDi and Text shaping.
    - Inline object/image embedding to the text flow.
    - Line/word breaking.
    - Fitting text to width and aligning to tab stops.
    - Input carets, selections and hit testing.
  - Implementations:
    - `TextServerFallback` module (simple text server without BiDi and shaping support).
    - `TextServerAdvanced` module (ICU/HarfBuzz/Graphite based text server), note: after `godot-cpp` catch up with the master, it can be converted to GDNative library with minimal changes (changing header names, build config, and adding wrapper from the demo repository - https://github.com/bruvzg/gdnative_text_server_template) and we'll need something like https://github.com/godotengine/godot-proposals/issues/831 to used it as library with the editor and project manager.
      - ICU optional dictionary data (cut down to 4 MB) can be embedded into the executable (default for `tools` build) or loaded for PCK (default for templates).
    - GDNative interface for text servers.
- `Font`
  - Font management moved to the `TextServer` to better handle shaping and font substitution.
  - `BitmapFont` and `DynamicFont` replaced with generic `Font`+`FontData`
  - Text can be drawn using LRU cached `font->draw_string`/`font->get_string_size` calls or by using `TextLine` / `TextParagraph` objects.
- Font size and outline color/size moved from base font to separate theme and draw functions options.
- Added interface mirroring for:
  - Child order in the containers.
  - Internal elements of complex controls (tree, itemlist, etc)
  - Added some mirrored icons and styles.
  - Interface mirroring can be overridden for each control, inherited or determined based on locale.
- Modified Windows and macOS input to handle UTF-16 input (multiple codepoints for one character).
- All controls modified to use new text rendering, line/word breaking and have additional properties added to control font features and text layout.
- BiDi override for handling specially structured text (URIs, paths, source code) added to `TextEdit`, `Label` and `LineEdit` (with optional used defined callback support for custom overrides).
- Added editor plugins for:
  - Font substitution control.
  - ICU data export.

<details>
  <summary>Screenshots</summary>

![new_ar](https://user-images.githubusercontent.com/7645683/94921367-e1980380-04c0-11eb-9bb0-554d21947903.png)
![new_he](https://user-images.githubusercontent.com/7645683/94921384-e78de480-04c0-11eb-8a9f-e35eed70ce92.png)
![new_hi](https://user-images.githubusercontent.com/7645683/94921398-ebba0200-04c0-11eb-9a2d-ec5374b584f9.png)
![new_th](https://user-images.githubusercontent.com/7645683/94921408-f1174c80-04c0-11eb-8e82-cfb027ee5360.png)
![new_ur](https://user-images.githubusercontent.com/7645683/94921417-f4123d00-04c0-11eb-890e-f7a7cb01a572.png)

</details>

I'll write a more detailed description with the use case samples for all changes in the PR in the next few days.

groud commented on core/math/math_defs.h:
```diff
@@ -75,7 +75,8 @@ enum HAlign {
 	HALIGN_CENTER,
-	HALIGN_RIGHT
+	HALIGN_RIGHT,
+	HALIGN_FILL,
```
What is this for? Are there languages that allow expanding the text?
I kind of find weird to have this here but not the same for VAlign.

groud commented on doc/classes/Button.xml:
```diff
@@ -57,9 +57,18 @@
 		<member name="text" type="String" setter="set_text" getter="get_text" default="&quot;&quot;">
 			The button's text that will be displayed inside the button's area.
 		</member>
+		<member name="text_direction" type="int" setter="set_text_direction" getter="get_text_direction" enum="Control.TextDirection" default="-1">
```
Shouldn't this be deduced from the font resource used? I believe it may be annoying to have to set manually no?

bruvzg replied:
It's deduced from locale normally (default value), this is for override (in case you need text in the other language in some controls) usually there's no need to change it. It can't be deduced from the font since most fonts for RTL languages have Latin characters as well (only big fontsets with the huge coverage, like Noto have separate files for each language).

groud replied:
Then I'd maybe rename it to text_direction_override. It would make it more obvious that this otherwise deduced automatically.

groud commented on doc/classes/CheckButton.xml:
```diff
@@ -58,16 +61,28 @@
-			The icon to display when the [CheckButton] is unchecked and disabled.
+			The icon to display when the [CheckButton] is unchecked and disabled (for left-to-right layouts).
+		</theme_item>
+		<theme_item name="off_disabled_mirrored" type="Texture2D">
```
Isn't this too many items? I kind of believe that simply mirroring the icon automatically would be enough in 99% of cases. I'd rather avoid defining a huge set of properties for all theme texture that could be flipped.

However, I believe we should maybe have a more generic approach here, likely allowing for a whole "mirrored" theme system. It would kind of inherit the base theme and replace only the items you want mirrored. That would avoid defining special properties for mirrored situations.

bruvzg replied:
Only few theme items need mirroring (option button and check button), and check button have text on it, simply mirroring texture won't work. Neither all icons need to be mirrored, only arrows and other things for forward/backward indication, rest is ok as is.

groud commented on doc/classes/CodeEdit.xml:
```diff
@@ -109,6 +110,8 @@
 		<member name="draw_line_numbers" type="bool" setter="set_draw_line_numbers" getter="is_draw_line_numbers_enabled" default="false">
 		</member>
+		<member name="layout_direction" type="int" setter="set_layout_direction" getter="get_layout_direction" override="true" enum="Control.LayoutDirection" default="2" />
+		<member name="text_direction" type="int" setter="set_text_direction" getter="get_text_direction" override="true" enum="Control.TextDirection" default="1" />
```
Same question as for buttons, can't this be deduced from the font used?

bruvzg commented:
## Changes to the Node API:

### `Control` changes:

---

1. Layout mirroring control (in `Control` and `Window`):
`is_layout_rtl()`, `layout_direction` property and `NOTIFICATION_LAYOUT_DIRECTION_CHANGED`
Layout mirrors the order of the controls in the container and internal components of the control, manually places controls are not moved, but additional layout dependent anchor presets are added `PRESET_xxx_BEGIN`/`PRESET_xxx_END` in addition to `PRESET_xxx_LEFT`/`PRESET_xxx_RIGHT`:

(top to bottom: LTR, RTL)
<img width="353" alt="Screenshot 2020-10-05 at 09 12 18" src="https://user-images.githubusercontent.com/7645683/95045790-70886400-06eb-11eb-9b7d-02bdd0f28f42.png">
<img width="370" alt="Screenshot 2020-10-05 at 09 12 45" src="https://user-images.githubusercontent.com/7645683/95045813-7aaa6280-06eb-11eb-8a23-04e1d95c645c.png">

For the testing purpose, layout can be force to `RTL` be setting `display/window/force_right_to_left_layout_direction` project setting.

`layout_direction` values:
- `LAYOUT_DIRECTION_INHERITED` - use layout direction of the parent control, or locale for root control (default).
- `LAYOUT_DIRECTION_LOCALE` - use locale to determine direction, to work with editor `Translation::get_tool_locale()` function is added (identical to `get_loacle()` for running projects).
- `TEXT_DIRECTION_LTR` and `TEXT_DIRECTION_RTL` - force specific direction, useful if you need fixed layout (e.g. buttons with directional arrows).

---

2. Text layout control (only for controls with text, or their elements, e.g. TreeItem ):
`text_direction` property -  controls base direction (default direction for not strongly directional text).
- `TEXT_DIRECTION_INHERITED` - use the same direction as layout direction (default).
- `TEXT_DIRECTION_AUTO` - use locale and text content (first strongly directional character) to determine direction.
- `TEXT_DIRECTION_LTR` and `TEXT_DIRECTION_RTL` - force specific default direction.

(Base direction can affect order of the elements in the mixed text: top-to-bottom Auto, LTR, RTL).
<img width="936" alt="Screenshot 2020-10-05 at 07 57 59" src="https://user-images.githubusercontent.com/7645683/95046470-e9d48680-06ec-11eb-9477-4956c49f158a.png">

In most cases base direction should be the same as the language of text (default behavior).

---

`structured_text_bidi_override` property and `_structured_text_parser` callback (only for controls with text):
BiDi algorithm is designed for natural text and in can't correctly display strings with a special structure (e.g URI, file path, regular expressions or source code):

![Screenshot 2020-10-05 at 09 06 27](https://user-images.githubusercontent.com/7645683/95046822-9c0c4e00-06ed-11eb-93ce-fdbb3cdd9874.png)

For example path for this directory structure will be displayed incorrectly (top string), structured text override add a callback function to split text into segments, then BiDi is applied to each of them individually (bottom string).

---

`language` property (only for controls with text):
Overrides locale for the control, and also controls language specific line breaking rules and shaping:

<img width="653" alt="Screenshot 2020-10-02 at 21 50 33" src="https://user-images.githubusercontent.com/7645683/95047195-556b2380-06ee-11eb-90fa-c822a5389b28.png">

---

`ot_features` property (only for controls with text):
Controls OpenType font features for the control: e.g. ligature types to use, number styles, small caps etc. (list of tags: https://docs.microsoft.com/en-us/typography/opentype/spec/featuretags)

<img width="829" alt="Screenshot 2020-10-02 at 21 48 31" src="https://user-images.githubusercontent.com/7645683/95047362-982cfb80-06ee-11eb-9757-e5466beae9ac.png">

---

`Control` and `Window` functions `add_theme_font_size_override`, `get_theme_font_size`, `has_theme_font_size`, `has_theme_font_size_override` and `Theme` functions: `clear_font_size`, `get_font_size`, `get_font_size_list`, `has_font_size`, `set_font_size`
Control font size in the similar manner to existing function for the `Font`, value `-1` is used as unset/default (same as `null` for the font).

---

`Label`: `outline_size` and `shadow_outline_size` - used to set outline size for the label.

---

`LineEdit` and `TextEdit`: `caret_mid_grapheme` disabled by default.
If enabled caret can be moved to each individual component of grapheme (e.g. diacritics and base character), if not moves over the whole grapheme.

(e.g in this case pressing delete will only remove diacritic, but pressing ← will jump over the full combined character)

![Screenshot 2020-10-05 at 07 59 06](https://user-images.githubusercontent.com/7645683/95047800-66686480-06ef-11eb-9b4a-24c461042e75.png)

(with mid caret enabled ← will jump to the position in between diacritic and the base character, caret will be displayed in the middle of combined character).

![Screenshot 2020-10-05 at 07 59 44](https://user-images.githubusercontent.com/7645683/95048028-c52dde00-06ef-11eb-983e-873f23db8d71.png)

I have encountered apps using both, but default behavior seems to be most common, not sure mid caret mode is useful at all.

---

`LineEdit` and `TextEdit`: `display_control` - if enabled control characters will be visible in the text (as special symbols if available in the font or hex codes if not), useful for more complex BiDi layout control. Also, popup menu to insert some of them as well as joiner characters to control line breaking added to the `LineEdit` and `TextEdit`.

bruvzg commented:
## `Font` and `FontData`:

Actual font implementation is move to text server, `Font` and `FontData` classes are used for easy access to TS font API.

### `FontData`:

Single font file (or memory buffer), handles both dynamic and bitmap fonts:

Additional language and script overrides can be set for each font data, these are used for font substitution in following cases:
- for the languages that are using the same script but different style: e.g. different Arabic languages use Naskh and Nastaliq styles and different fonts but it's the same script (writing system), used by editor to handle Urdu and Persian locales.
- script overrides can be useful for, rare languages and big font sets to select proper font faster (TrueType fonts have flags only for some scripts, if not selected manually shaped will try every font in the list instead of going directly to the correct one).
- script overrides can be used for the fonts with incomplete character sets, e.g many Latin fonts have limited set of Greek characters for scientific text, that is not enough for real Greek text, but such font will have flag for Greek support. In this case you may want to add override to avoid using it for Greek and add separate font.

Note: Shaped will try all fonts until all character are found, overrides only affect search order.

Note: there was `distance_field_hint` property in the BitmapFont that seems to be not implemented, I have preserved it as is in new `FontData` as is.

`FontData` have `base_size` used to load on font creation and which is used for draw function with `p_size == -1`, but it is capable to load and use any font size on demand.

### `Font`:
Font "fallback"/substitution uses single list of the `FontData` in the similar manner `DynamicFont` was, but there's no primary one since order is determined by language/script support and overrides.

`draw/draw_haling` Function merged to the single one, and multiline draw function was added, each of of this have `size` function with the same API and use LRU cache to reuse shaped data.

Additionally, `core/math/math_defs.h:: HALIGN_FILL` is added for use of haligned draw.

`CanvasItem` text draw functions are changed in the same way.

If font size is set to `-1`, base_size is used.

```
	Size2 get_string_size(const String &p_text, int p_size = -1) const;
	Size2 get_multiline_string_size(const String &p_text, float p_width = -1, int p_size = -1, uint8_t p_flags = TextServer::BREAK_MANDATORY | TextServer::BREAK_WORD_BOUND) const;

	void draw_string(RID p_canvas_item, const Point2 &p_pos, const String &p_text, HAlign p_align = HALIGN_LEFT, float p_width = -1, int p_size = -1, const Color &p_modulate = Color(1, 1, 1), int p_outline_size = 0, const Color &p_outline_modulate = Color(1, 1, 1, 0), uint8_t p_flags = TextServer::JUSTIFICATION_KASHIDA | TextServer::JUSTIFICATION_WORD_BOUND) const;
	void draw_multiline_string(RID p_canvas_item, const Point2 &p_pos, const String &p_text, HAlign p_align = HALIGN_LEFT, float p_width = -1, int p_max_lines = -1, int p_size = -1, const Color &p_modulate = Color(1, 1, 1), int p_outline_size = 0, const Color &p_outline_modulate = Color(1, 1, 1, 0), uint8_t p_flags = TextServer::BREAK_MANDATORY | TextServer::BREAK_WORD_BOUND | TextServer::JUSTIFICATION_KASHIDA | TextServer::JUSTIFICATION_WORD_BOUND) const;
```

Char draw functions are preserved (for drawing single chars, numbers and other simple stuff), these functions do not use `TextServer` shaping or BiDi and draw chars directly:
```
	Size2 get_char_size(char32_t p_char, char32_t p_next = 0, int p_size = -1) const;
	float draw_char(RID p_canvas_item, const Point2 &p_pos, char32_t p_char, char32_t p_next = 0, int p_size = -1, const Color &p_modulate = Color(1, 1, 1), int p_outline_size = 0, const Color &p_outline_modulate = Color(1, 1, 1, 0)) const;
```

Note: font functions like `get_ascent/get_descent` return average of the `FontData` ascents/descents, in the real text, character elements can be outside of font ascent and descent and these functions are only useful as rough estimate (e.g. for empty text field size).

For the static texts, local (control specific) cache, and more precise control `TextLine` and `TextPragraph` classes are added as the easy-to-use wrapper over TS API.

groud commented:
> Layout mirrors the order of the controls in the container and internal components of the control, manually places controls are not moved, but additional layout dependent anchor presets are added PRESET_xxx_BEGIN/PRESET_xxx_END in addition to PRESET_xxx_LEFT/PRESET_xxx_RIGHT:

I am not really sure this is the best implementation. Those are presets used by the `set_anchors_and_margins` presets, which is basically a helper to call several `set_anchor` and `set_margin`. As this one is a helper, it kind of annoys me to allow RTL feature only in the preset setting function, but not in the normal calls to `set_anchor` and `set_margin`.

I would eventually suggest to instead add an argument to all those function, to switch a to a RTL layout. So for `set_margin`, `set_margins_preset`, `set_anchor`, `set_anchors_preset`, `set_anchor_and_margin` and `set_anchors_and_margins_preset` I would add an argument to switch to the opposite layout. We could also rename (or create an alias for compatibility reasons) the MARGIN_LEFT and MARGIN_RIGHT contants to MARGIN_HBEGIN and MARGIN_HEND (or somehow like that).

bruvzg commented:
## `TextServer` API:

TextServer provides the following APIs:

`font` functions for Font creation and control, a backend for `FontData`.
`shaped_text` functions for text buffers, a backend for `Font` draw/size functions and `TextLine`, `TextPragraph`:

Shaped text buffer APIs are lazy and won't do any processing until output data is requested, substring will reuse shaping results of the parent string until they are changed.

```
bool shaped_text_add_string(RID p_shaped, const String &p_text, const Vector<RID> &p_fonts, int p_size, const String &p_ot_features = "", const String &p_language = "");
bool shaped_text_add_object(RID p_shaped, Variant p_key, const Size2 &p_size, VAlign p_inline_align = VALIGN_CENTER);
```
Functions are used to fill up the buffer with segments of text and inlined objects (e.g. images and tables in RichTextLabel, object placement are affected by BiDi and sizes used during line breaking).

Text server provides functions for breaking text into lines and words (can use optionally use ICU data dictionaries for better breaking) and extraction of substrings (without reshaping).

Also, it has helper functions for input controls:
- caret handling: when caret is in between the text segments with different directionally, it can be displayed in multiple locations in the same time (input will appear in the different places of text depending on the inputted text direction). As well as caret movement over grapheme (see. mid gr. caret above.)
<img width="181" alt="Screenshot 2020-10-05 at 10 45 43" src="https://user-images.githubusercontent.com/7645683/95052638-f3afb700-06f7-11eb-98bd-86962d0d08b6.png">
- getting selection rectangles, since selection in the bidirectional text is not always continues and characters can overlap, it's easier to get it in one go.
- hit testing for caret placement.

---

Text buffer workflow looks as following:

1. Buffer is filled with text spans and objects (for the rest of processing it's just single "Object replacement character" with the specified size) and its properties are set. (at this stage no processing is done, input data is stored as is).

2. As soon as any output is requested (e.g. get_size, draw, or get_glyphs is called).
- Text is split to the segments (runs) based on the script used (custom iterator based on ICU Unicode data functions).
- ICU BiDi is used to detect directions and order of the text segments (in case structured text override is active it's used multiple times), BiDi context is saved to be reused with substrings.
- Each segment of the text with the uniform script, direction, language and font (language and font are taken from the spans) is shaped (characters are transformed to the font specific glyphs and offset of the glyphs are determined) using HarfBuzz, in the order returned by BiDi. FontData order is determined based on script and language, shaper will try to use each one until all subruns are shaped successfully or there are no more fonts in the list. Failed characters are stored as is and later displayed as the hexcodes when string is drawn.
As the result all glyphs are stored to the vector in the order in which they are displayed.

3. As soon as any of the line breaking functions is called.
- ICU break iterator is used to determine line and word breaks, in case there's no space character between the words, virtual zero width glyph is added, flags for breaks are added to existing glyphs.
- Glyphs are sorted in the logical order (as their source characters were in the original string).
- Sorted glyph widths and flags are used to determine line breaks.

4. As soon as fit to wight is called:
- Word breaking is done if it was not called by line breaking.
- Kashidas are detected in word, or added to the word in appropriate places.
- First kashidas and then spaces are extended to fit the text.

(top-to-bottom: kashida and space justification, no justification, space justification)
<img width="523" alt="Screenshot 2020-10-05 at 11 35 40" src="https://user-images.githubusercontent.com/7645683/95057588-10032200-06ff-11eb-9d3b-f16cea64fa8e.png">

Breaks and kashidas are reused for the substring alignment without repeating the step.

bruvzg commented:
> I would eventually suggest to instead add an argument to all those function, to switch a to a RTL layout. 

This might be better solution indeed.

bruvzg commented:
Note: Rebased and temporary re-added `env.AppendUnique(CCFLAGS=["/utf-8"])` flag that was removed in #42634 (will rebase again after #42668).

There are multiple UTF-8 constant strings used in the tests and in the editor (cause tests to fail) and as well as in ICU (cause build failure).

bruvzg commented:
Rebased and replace preset `PRESET_xxx_LEFT/PRESET_xxx_RIGHT` with full anchor/margin mirroring.

bruvzg commented:
Demos for the main new features: https://github.com/godotengine/godot-demo-projects/pull/538

bruvzg commented:
Rebased, added option to enable/disable coding ligatures (contextual alternates) in the built-in code editor.

khaledhosny commented on editor/code_editor.cpp:
```diff
@@ -1473,6 +1489,12 @@ void CodeTextEditor::_on_settings_change() {
 
 	font_size = EditorSettings::get_singleton()->get("interface/editor/code_font_size");
 
+	if (bool(EditorSettings::get_singleton()->get("interface/editor/code_font_ligatures"))) {
```
That is a very specific name for a general feature. `calt` might not be used for ligatures (an Arabic or Indic monospaced font for instance), and ”coding ligatures” might use other features like `liga`. Either make the name reflect what the code is actually doing (not the its intent) or make it a general OpenType features control.

Calinou replied:
> or make it a general OpenType features control.

The Editor Settings dialog can only expose simple booleans, integers, floats or "fixed" enums, not an enum that's dynamically constructed based on the currently loaded font.

khaledhosny replied:
A fixed `enum` should work, the [number of OpenType features](https://docs.microsoft.com/en-us/typography/opentype/spec/featurelist) is finite, and though the list is long, many of the features are required (for proper text layout) and should not be controllable.

bruvzg replied:
Fixed enum won't give full control, since some features can be set to the integer value not just enabled/disabled (swashes and other stylistic features), and each feature can be applied to the sub range of the string.

For generic controls, OT features are currently set as the single string property (e.g `swsh=2,lnum,zero`), but for the code editor it's probably better to have some fixed list too, maybe:

- `code_font_ligatures` → (enum) "Default, Disable ligatures, Use custom OT features"
- `code_font_ot_features` → (string) for full control, used if "custom" is selected.

khaledhosny replied:
> `code_font_ligatures` → (enum) "Default, Disable ligatures, Use custom OT features"

My point is that a name like `code_font_ligatures` is misleading, if you really want to control `calt` (though I’m pretty sure there are ”coding fonts” using `liga`), just call it what it is; contextual ligatures.

bruvzg replied:
At least "Fira Code", "Iosevka" and "JetBrains Mono" use `calt` to control coding ligatures `liga` do not affect it at all.

bruvzg replied:
I guess calling it `code_font_contextual_ligatures` instead of `code_font_ligatures` won't hurt, it's still easy to find by searching for "ligatures", but on the other hand `calt` is "contextual alternates" not "contextual ligatures" (`clig` which do not have any effect on coding ligatures).

bruvzg replied:
Additionally, some coding fonts seem to have non-standard features for specific subsets of ligatures, like `HSKL` for Haskell ligatures and `CLIK` for C ligatures.

bruvzg replied:
Changed it to `Default, Disable coding ligatures, Use custom OpenType feature set` enum + custom features string.

<img width="864" alt="Screenshot 2020-10-18 at 22 19 15" src="https://user-images.githubusercontent.com/7645683/96377726-32318100-1190-11eb-9cde-cf9520ae9694.png">

khaledhosny replied:
Great, thanks. (`Ot` should probably be `OT` or better `OpenType`).

bruvzg replied:
> Ot should probably be OT or better OpenType

It should be, but currently editor setting names are automatically converted from snake case to title case, AFAIK there's no way to archive this (there are many other setting that should be all upper case abbreviations but are not for the same reason).

Calinou replied:
Naming it `Opentype` would be better. (It's all we can do with the current automatic capitalization; we'd need https://github.com/godotengine/godot/pull/32734 to be merged to improve this.)

bruvzg replied:
Renamed `ot` to `opentype` (editor setting and generic control properties), and added simple editor inspector plugin to edit features:
<img width="467" alt="Screenshot 2020-10-19 at 12 13 18" src="https://user-images.githubusercontent.com/7645683/96425567-d3f2b580-1204-11eb-9f7e-7c9c221f7f85.png">

bruvzg commented:
Some docs for better understanding of what "base direction" is:
https://www.w3.org/International/articles/inline-bidi-markup/uba-basics
https://www.w3.org/International/questions/qa-bidi-unicode-controls

And Godot demo project with the same samples: [W3.zip](https://github.com/godotengine/godot/files/5479893/W3.zip)

<img width="788" alt="Screenshot 2020-11-03 at 09 59 22" src="https://user-images.githubusercontent.com/7645683/97961070-434ad680-1dbb-11eb-89d9-ef9947ac168f.png">

Paulb23 reviewed (commented):
Did a first pass through GUI controls, nothing major jumped out, noted a couple small things inline. 

One question around `CodeEdit`, as it's used for editing `TextFiles` would it be worth detecting the direction rather then hardcoding it?


<details><summary>Some bugs</summary>

- Caret at the end of a wrapped line is also drawn at the start of the next
![image](https://user-images.githubusercontent.com/6584330/98443668-98148700-2104-11eb-93d1-305ea492374b.png)

- Not wrapping early enough:
![image](https://user-images.githubusercontent.com/6584330/98443690-d4e07e00-2104-11eb-8071-c4e4326a790b.png)

- Word highlight doesn't handle wrapped lines:
![image](https://user-images.githubusercontent.com/6584330/98443716-0ce7c100-2105-11eb-9bd6-366f78e74d36.png)

- In Insert mode the caret underlines the entire tab, before it used to highlight a the equivalent of one char, also vertical when at the end of line.

- After changing zoom with `ctrl+mouse wheel`, `ctrl+0` doesn't always reset the font size.

</details>

bruvzg commented on scene/gui/code_edit.cpp:
```diff
@@ -368,7 +370,9 @@ void CodeEdit::_lines_edited_from(int p_from_line, int p_to_line) {
 		line_number_digits++;
 	}
-	set_gutter_width(line_number_gutter, (line_number_digits + 1) * cache.font->get_char_size('0').width);
+	set_gutter_width(main_gutter, get_row_height());
```
> Not sure I understand this change

Was added as a workaround, seems like it's not necessary now. Removed.

> Caret at the end of a wrapped line is also drawn at the start of the next

Fixed

> Not wrapping early enough:

Fixed (was not taking tabs into account)

> Word highlight doesn't handle wrapped lines:

Fixed

> In Insert mode the caret underlines the entire tab, before it used to highlight a the equivalent of one char

What's wrong with this? It's always using actual width of the char when possible (including tabs). At least some other editors do the same (vscode).

> also vertical when at the end of line.

Fixed

> ctrl+0 doesn't always reset the font size.

Fixed

akien-mga replied:
>> In Insert mode the caret underlines the entire tab, before it used to highlight a the equivalent of one char
>
> What's wrong with this? It's always using actual width of the char when possible (including tabs). At least some other editors do the same (vscode).

See also this related issue #42656 and PR #42975.

Paulb23 replied:
> What's wrong with this? It's always using actual width of the char when possible (including tabs). At least some other editors do the same (vscode).

Seems like this is editor independent, Sublime does't do this. QT Creator and Intellij also don't, but they use a block caret for overwrite mode rather then underlining it. I personally prefer underling the equivalent of one char for tabs, but given there is no real "standard", should be okay to keep as is for now.

bruvzg commented on modules/text_server_adv/text_server_adv.cpp:
```diff
@@ -0,0 +1,2152 @@
+	{ "dz", U"༠༡༢༣༤༥༦༧༨༩.", U"%", U"e" },
+	{ "sat", U"᱐᱑᱒᱓᱔᱕᱖᱗᱘᱙.", U"%", U"e" },
+	{ "my", U"၀၁၂၃၄၅၆၇၈၉.", U"%", U"e" },
+	{ String(), String(), String() },
```
Fixed

bruvzg commented on scene/gui/line_edit.h:
```diff
@@ -185,19 +215,46 @@ class LineEdit : public Control {
 	void append_at_cursor(String p_text);
 	void clear();
 
+	void set_mid_grapheme_cursor(const bool p_enabled);
```
Renamed to `*_mid_grapheme_caret_enabled`

bruvzg commented on scene/gui/line_edit.h:
```diff
@@ -185,19 +215,46 @@ class LineEdit : public Control {
+	void set_language(const String &p_language);
+	String get_language() const;
+
+	void set_display_control(bool p_display_control);
```
Renamed to `*_draw_control_chars`

bruvzg commented on scene/gui/text_edit.cpp:
```diff
@@ -341,6 +381,9 @@ void TextEdit::_click_selection_held() {
 	Point2 mp = get_local_mouse_position();
+	if (is_layout_rtl()) {
+		mp.x = get_size().width - mp.x;
+	}
```
Moved to `_get_local_mouse_pos()`, this code was used 7 time for the reference in TextEdit.

bruvzg commented:
Rebase after #42841, and added C# code samples to the new Font docs.

akien-mga commented:
I got a crash trying to assign `DejaVuSans.ttf` as FontData for a Font used in a Label with some Farsi text:
```
[1] /lib64/libc.so.6(+0x3a3a0) [0x7f03b32863a0] (??:0)
[2] godot-git() [0x52e6a19] (/home/akien/Projects/godot/godot.git/thirdparty/freetype/src/base/ftstream.c:800)
[3] godot-git() [0x536738e] (/home/akien/Projects/godot/godot.git/thirdparty/freetype/src/sfnt/ttload.c:206)
[4] godot-git() [0x53675f8] (/home/akien/Projects/godot/godot.git/thirdparty/freetype/src/sfnt/ttload.c:390)
[5] godot-git() [0x53566b8] (/home/akien/Projects/godot/godot.git/thirdparty/freetype/src/sfnt/sfobjs.c:571)
[6] godot-git() [0x5342758] (/home/akien/Projects/godot/godot.git/thirdparty/freetype/src/truetype/ttobjs.c:635)
[7] godot-git() [0x52dca6e] (/home/akien/Projects/godot/godot.git/thirdparty/freetype/src/base/ftobjs.c:1421)
[8] godot-git() [0x52de295] (/home/akien/Projects/godot/godot.git/thirdparty/freetype/src/base/ftobjs.c:2493)
[9] godot-git(FT_Open_Face+0x36) [0x52de000] (/home/akien/Projects/godot/godot.git/thirdparty/freetype/src/base/ftobjs.c:2385)
[10] DynamicFontDataAdvanced::get_data_for_size(int, int) (/home/akien/Projects/godot/godot.git/modules/text_server_adv/dynamic_font_adv.cpp:103)
[11] DynamicFontDataAdvanced::get_hb_handle(int) (/home/akien/Projects/godot/godot.git/modules/text_server_adv/dynamic_font_adv.cpp:820)
[12] TextServerAdvanced::_shape_run(TextServerAdvanced::ShapedTextDataAdvanced*, int, int, hb_script_t, hb_direction_t, Vector<RID>, int, int) (/home/akien/Projects/godot/godot.git/modules/text_server_adv/text_server_adv.cpp:1561)
[13] TextServerAdvanced::shaped_text_shape(RID) (/home/akien/Projects/godot/godot.git/modules/text_server_adv/text_server_adv.cpp:1885 (discriminator 19))
[14] TextServerAdvanced::shaped_text_update_breaks(RID) (/home/akien/Projects/godot/godot.git/modules/text_server_adv/text_server_adv.cpp:1216)
[15] TextServer::shaped_text_get_line_breaks(RID, float, int, unsigned char) const (/home/akien/Projects/godot/godot.git/servers/text_server.cpp:610)
[16] Label::_shape() (/home/akien/Projects/godot/godot.git/scene/gui/label.cpp:103)
[17] Label::get_minimum_size() const (/home/akien/Projects/godot/godot.git/scene/gui/label.cpp:361)
[18] Control::_update_minimum_size_cache() (/home/akien/Projects/godot/godot.git/scene/gui/control.cpp:179)
[19] Control::get_combined_minimum_size() const (/home/akien/Projects/godot/godot.git/scene/gui/control.cpp:200)
[20] Control::_update_minimum_size() (/home/akien/Projects/godot/godot.git/scene/gui/control.cpp:297)
[21] void call_with_variant_args_helper<__UnexistingClass>(__UnexistingClass*, void (__UnexistingClass::*)(), Variant const**, Callable::CallError&, IndexSequence<>) (/home/akien/Projects/godot/godot.git/./core/variant/binder_common.h:217 (discriminator 4))
[22] void call_with_variant_args_dv<__UnexistingClass>(__UnexistingClass*, void (__UnexistingClass::*)(), Variant const**, int, Callable::CallError&, Vector<Variant> const&) (/home/akien/Projects/godot/godot.git/./core/variant/binder_common.h:334)
[23] MethodBindT<>::call(Object*, Variant const**, int, Callable::CallError&) (/home/akien/Projects/godot/godot.git/./core/object/method_bind.h:288)
[24] Object::call(StringName const&, Variant const**, int, Callable::CallError&) (/home/akien/Projects/godot/godot.git/core/object/object.cpp:787 (discriminator 1))
[25] Callable::call(Variant const**, int, Variant&, Callable::CallError&) const (/home/akien/Projects/godot/godot.git/core/variant/callable.cpp:53)
[26] MessageQueue::_call_function(Callable const&, Variant const*, int, bool) (/home/akien/Projects/godot/godot.git/core/object/message_queue.cpp:259)
[27] MessageQueue::flush() (/home/akien/Projects/godot/godot.git/core/object/message_queue.cpp:306)
[28] SceneTree::iteration(float) (/home/akien/Projects/godot/godot.git/scene/main/scene_tree.cpp:415 (discriminator 2))
[29] Main::iteration() (/home/akien/Projects/godot/godot.git/main/main.cpp:2481)
[30] OS_LinuxBSD::run() (/home/akien/Projects/godot/godot.git/platform/linuxbsd/os_linuxbsd.cpp:261)
[31] godot-git(main+0x135) [0x1fa2ce7] (/home/akien/Projects/godot/godot.git/platform/linuxbsd/godot_linuxbsd.cpp:60)
[32] /lib64/libc.so.6(__libc_start_main+0xea) [0x7f03b3272d1a] (??:0)
[33] godot-git(_start+0x2a) [0x1fa2b0a] (??:?)
-- END OF BACKTRACE --
```

Here's my test project: [Mahi.zip](https://github.com/godotengine/godot/files/5523760/Mahi.zip)

This project was created when I last tested this PR in Oct 14, and there the Label seemed show Farsi fine without having to set a custom font, though I might be misrembering. Anyway, I copied DejaVuSans.ttf today to try to fix it with the current version of the PR (I don't know if it has necessary glyphs, but it was just a test) and got this crash when assigning it.

akien-mga commented:
> *Edit:* I'm not getting a crash on a brand new project when creating a Label with text `ماهی سیاه کوچولو` and assigning `DejaVuSans.ttf` as the font (which does have needed glyphs as I see now).

In that new test project, the root node was Control, so I tried to change it to Node2D as in my other project.

Then I tried to move the Label and got a new crash:
```
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] /lib64/libc.so.6(+0x3a3a0) [0x7f2f2bcbd3a0] (??:0)
[2] /lib64/libc.so.6(__libc_malloc+0xcc) [0x7f2f2bd08cbc] (??:0)
[3] Memory::alloc_static(unsigned long, bool) (/home/akien/Projects/godot/godot.git/core/os/memory.cpp:76 (discriminator 4))
[4] DynamicFontDataAdvanced::is_script_supported(unsigned int) const (/home/akien/Projects/godot/godot.git/modules/text_server_adv/dynamic_font_adv.cpp:569)
[5] TextServerAdvanced::font_is_script_supported(RID, String const&) const (/home/akien/Projects/godot/godot.git/modules/text_server_adv/text_server_adv.cpp:435 (discriminator 2))
[6] FontData::is_script_supported(String const&) const (/home/akien/Projects/godot/godot.git/scene/resources/font.cpp:357)
[7] EditorFontPreviewPlugin::generate_from_path(String const&, Vector2 const&) const (/home/akien/Projects/godot/godot.git/editor/plugins/editor_preview_plugins.cpp:866 (discriminator 1))
[8] EditorFontPreviewPlugin::generate(Ref<Resource> const&, Vector2 const&) const (/home/akien/Projects/godot/godot.git/editor/plugins/editor_preview_plugins.cpp:917)
[9] EditorResourcePreview::_generate_preview(Ref<ImageTexture>&, Ref<ImageTexture>&, EditorResourcePreview::QueueItem const&, String const&) (/home/akien/Projects/godot/godot.git/editor/editor_resource_preview.cpp:158 (discriminator 3))
[10] EditorResourcePreview::_thread() (/home/akien/Projects/godot/godot.git/editor/editor_resource_preview.cpp:238 (discriminator 3))
DisplayServer::_create_window 7 want rect: 702, 498, 516, 83 got rect 702, 498, 516, 83
early out of data
[11] EditorResourcePreview::_thread_func(void*) (/home/akien/Projects/godot/godot.git/editor/editor_resource_preview.cpp:100)
[12] ThreadPosix::thread_callback(void*) (/home/akien/Projects/godot/godot.git/drivers/unix/thread_posix.cpp:73)
[13] /lib64/libpthread.so.0(+0x8df0) [0x7f2f2c177df0] (??:0)
[14] /lib64/libc.so.6(clone+0x3f) [0x7f2f2bd790ef] (??:0)
-- END OF BACKTRACE --
```

bruvzg commented:
> re-assigned the DejaVuSans.ttf 

Seems like something wrong with assigning file as resource directly, creating `FontData` instance and setting path works fine, assigning file as `FontData` crashes.

bruvzg commented:
Rebased, crash should be fixed.

akien-mga commented:
Thanks, I confirm that both crashes seem fixed (at least I couldn't reproduce them).

I found this issue where strings without translation (i.e. using the source English string) are still treated as RTL (see "...Install Android Build Template").
![Screenshot_20201111_153732](https://user-images.githubusercontent.com/4701338/98824698-e4790300-2433-11eb-90d9-e53b4a91ca25.png)

I guess this can be done as a second step once this is merged, but we'll likely need UX feedback from native speakers to see how to configure UI mirroring for different parts of the editor, as e.g. in the mostly English Editor Settings (as settings sections and names can't be localized), this can be confusing IMO:
![Screenshot_20201111_154056](https://user-images.githubusercontent.com/4701338/98825020-4b96b780-2434-11eb-9ac8-d6f7f67bf809.png)

But I'm just mentioning it for reference, I do think that it's better to handle this editor-specific configuration work in follow up PRs and keep this one focused on the feature implementation.

bruvzg commented:
> I found this issue where strings without translation (i.e. using the source English string) are still treated as RTL (see "...Install Android Build Template").

Bug in the PopupMenu, the default direction was set to Inherited instead of Auto.

> but we'll likely need UX feedback from native speakers to see how to configure UI mirroring for different parts of the editor

Definitely need feedback from native speakers, I have no idea what's the best way to handle it (we can easily set some parts of the editor to be always LTR if it's desired).

> as settings sections and names can't be localized

What's the reason for not translating editor/project settings and property names?

akien-mga commented:
> What's the reason for not translating editor/project settings and property names?

Mostly that those settings are generated directly from their identifier (`some_section/some_setting` becomes "Some Section / Some Settings"), and if we translate them we introduce some additional friction:
- The visual setting name will no longer match the identifier.
- It will also not match what is being described in the class reference.
- When you add new settings yourself in Project Settings, you'd need to be able to provide both an identifier that can match existing categories if you wish so, and a translation for the sections and name substrings.

Initially it was considered simpler to just keep it automated and thus in English. We could reconsider this if we find ways to alleviate the issues, but it would require a significant rethinking of the UI to accommodate it without losing the flexibility of the current automated system.

bruvzg commented:
Note: #28537 is still relevant for the new `FontData`.

When a font file is assigned as `FontData` directly (saved in the parent `.tres` as "ext_resource") none of the properties is saved, when new `FontData` is instanced first (saved as separate `.tres` or "sub_resource" in the parent `.tres`) properties are saved/loaded correctly.

bruvzg commented:
Updated OpenType features (and variations in #43030) to use individual getters/setters instead of a single string property, and added editor inspector plugin to config it (readable names for standard features, only supported features shown).

<img width="825" alt="opentype_features" src="https://user-images.githubusercontent.com/7645683/98920843-42abf180-24d9-11eb-9f86-1b15ada04c18.png">

Added RTL control to the demo (https://github.com/godotengine/godot-demo-projects/pull/538) to show setting OpenType features for the span of text.

![font_features](https://user-images.githubusercontent.com/7645683/98921415-f0b79b80-24d9-11eb-8896-f7a4b664cfef.png)

akien-mga commented:
Did another test of this PR today and got a crash again when assigning a `FontData` resource using `Load` for an in-use custom `Font` for a `Label`:

- Added Label
- Copied `Test আমি খেলছি` as text
- Add custom font in inspector with "New Font"
- Expand "Data", use the "Load" option to pick `DejaVuSans.ttf`
- Crash

```
ERROR: Texture is not initialized.
   at: update (scene/resources/texture.cpp:178)

Thread 1 "godot-git" received signal SIGSEGV, Segmentation fault.
0x000000000535c11f in af_glyph_hints_reload (hints=0x7fffffff9040, outline=0x23690e78) at thirdparty/freetype/src/autofit/afhints.c:1042
1042                out_x += next->fx - point->fx;
(gdb) bt
#0  0x000000000535c11f in af_glyph_hints_reload (hints=0x7fffffff9040, outline=0x23690e78) at thirdparty/freetype/src/autofit/afhints.c:1042
#1  0x0000000005362dad in af_latin_hints_apply (glyph_index=72, hints=0x7fffffff9040, outline=0x23690e78, metrics=0x2514f3b0) at thirdparty/freetype/src/autofit/aflatin.c:3544
#2  0x0000000005363980 in af_loader_load_glyph (loader=0x7fffffff8fc0, module=0x237d9d30, face=0xac66a60, glyph_index=72, load_flags=75777) at thirdparty/freetype/src/autofit/afloader.c:425
#3  0x0000000005364e2b in af_autofitter_load_glyph (module=0x237d9d30, slot=0x2322c2f0, size=0x23854980, glyph_index=72, load_flags=65536) at thirdparty/freetype/src/autofit/afmodule.c:535
#4  0x00000000052c66b5 in FT_Load_Glyph (face=0xac66a60, glyph_index=72, load_flags=65536) at thirdparty/freetype/src/base/ftobjs.c:939
#5  0x00000000024df11c in DynamicFontDataAdvanced::update_glyph (this=0xabf46b0, p_size=16, p_index=72) at modules/text_server_adv/dynamic_font_adv.cpp:446
#6  0x00000000024de48a in DynamicFontDataAdvanced::draw_glyph (this=0xabf46b0, p_canvas=..., p_size=16, p_pos=..., p_index=72, p_color=...) at modules/text_server_adv/dynamic_font_adv.cpp:890
#7  0x00000000024ba1c5 in TextServerAdvanced::font_draw_glyph (this=0x91ea7b0, p_font=..., p_canvas=..., p_size=16, p_pos=..., p_index=72, p_color=...) at modules/text_server_adv/text_server_adv.cpp:774
#8  0x0000000003ed2737 in Label::_notification (this=0x238c5a50, p_what=30) at scene/gui/label.cpp:324
#9  0x0000000003db76c6 in Label::_notificationv (this=0x238c5a50, p_notification=30, p_reversed=false) at ./scene/gui/label.h:37
#10 0x00000000051a1d87 in Object::notification (this=0x238c5a50, p_notification=30, p_reversed=false) at core/object/object.cpp:793
#11 0x0000000003d07f37 in CanvasItem::_update_callback (this=0x238c5a50) at scene/main/canvas_item.cpp:434
#12 0x0000000002060735 in call_with_variant_args_helper<__UnexistingClass>(__UnexistingClass*, void (__UnexistingClass::*)(), Variant const**, Callable::CallError&, IndexSequence<>) (p_instance=0x238c5a50, 
    p_method=(void (__UnexistingClass::*)(__UnexistingClass * const)) 0x3d07e48 <CanvasItem::_update_callback()>, p_args=0x7fffffffcae8, r_error=...) at ./core/variant/binder_common.h:212
#13 0x000000000205ffb8 in call_with_variant_args_dv<__UnexistingClass> (p_instance=0x238c5a50, p_method=(void (__UnexistingClass::*)(__UnexistingClass * const)) 0x3d07e48 <CanvasItem::_update_callback()>, 
    p_args=0x0, p_argcount=0, r_error=..., default_values=...) at ./core/variant/binder_common.h:333
#14 0x000000000205f100 in MethodBindT<>::call(Object*, Variant const**, int, Callable::CallError&) (this=0x9f530e0, p_object=0x238c5a50, p_args=0x0, p_arg_count=0, r_error=...)
    at ./core/object/method_bind.h:287
#15 0x00000000051a1c6a in Object::call (this=0x238c5a50, p_method=..., p_args=0x0, p_argcount=0, r_error=...) at core/object/object.cpp:784
#16 0x0000000004f798d1 in Callable::call (this=0x7ffff596e698, p_arguments=0x0, p_argcount=0, r_return_value=..., r_call_error=...) at core/variant/callable.cpp:53
#17 0x000000000519aff4 in MessageQueue::_call_function (this=0x91d3e60, p_callable=..., p_args=0x7ffff596e6b0, p_argcount=0, p_show_error=false) at core/object/message_queue.cpp:258
#18 0x000000000519b30d in MessageQueue::flush (this=0x91d3e60) at core/object/message_queue.cpp:304
#19 0x0000000003d752a9 in SceneTree::iteration (this=0x9a3c170, p_time=0.0166666675) at scene/main/scene_tree.cpp:414
#20 0x0000000001f616fd in Main::iteration () at main/main.cpp:2472
#21 0x0000000001f2a1a8 in OS_LinuxBSD::run (this=0x7fffffffd150) at platform/linuxbsd/os_linuxbsd.cpp:261
#22 0x0000000001f27c07 in main (argc=2, argv=0x7fffffffd658) at platform/linuxbsd/godot_linuxbsd.cpp:58
```

bruvzg commented:
> - Expand "Data", use the "Load" option to pick DejaVuSans.ttf
> - Crash

It was caused by `EditorFontPreviewPlugin::generate_from_path` drawing a sample string (and updating font glyphs) from the separate thread simultaneously to the Label update (seems like it's the same crash as the previous one, but I did not notice it was running in a thread first time).
Changed TS and Font to be `_THREAD_SAFE_CLASS_` classes.

reduz commented:
This is an incredible piece of work. I think its really good, probably best to leave it up to the community to use and give feedback on usage but I think it should be good enough. My main fear is backwards compatibility with older scenes, but I guess it should work?

reduz commented:
We are discussing a way to do compatibility filters between older and newer Godot versions, so there might be a way I guess.

akien-mga commented:
Here we go! A huge thank you for this amazing feature!

aaronfranke commented:
```cpp
bool Control::is_layout_rtl() const {
	if (data.layout_dir == LAYOUT_DIRECTION_INHERITED) {
		Window *parent_window = Object::cast_to<Window>(get_parent());
		Control *parent_control = get_parent_control();
		if (parent_control) {
			return parent_control->is_layout_rtl();
		} else if (parent_window) {
			return parent_window->is_layout_rtl();
		} else {
			if (GLOBAL_GET("display/window/force_right_to_left_layout_direction")) {
				return true;
			}
```

It seems that this can fail in the project manager. `GLOBAL_GET` will fail in the project manager because there is no `ProjectSettings` singleton, and `Control::is_layout_rtl` is used in the project manager. So if `data.layout_dir == LAYOUT_DIRECTION_INHERITED` and neither `parent_control` or `parent_window` exist, the project manager will crash.

```
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x46210) [0x7f92652e4210] (??:0)                                                                          
[2] Object::get(StringName const&, bool*) const (/home/aaronfranke/workspace/godot/core/object/object.cpp:445)
[3] Control::is_layout_rtl() const (/home/aaronfranke/workspace/godot/scene/gui/control.cpp:442 (discriminator 4))
[4] BoxContainer::_resort() (/home/aaronfranke/workspace/godot/scene/gui/box_container.cpp:47 (discriminator 2))
[5] BoxContainer::set_alignment(BoxContainer::AlignMode) (/home/aaronfranke/workspace/godot/scene/gui/box_container.cpp:310)
```

KoBeWi commented:
So I noticed it just now, but why does this PR introduce theme font sizes instead of using theme constants for this purpose?

bruvzg commented:
> So I noticed it just now, but why does this PR introduce theme font sizes instead of using theme constants for this purpose?

Font size have some extra fallback logic for the Theme and global default values, constants use `0` when net set.
